src/runtime/storageNG/backing-storage-proxy.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * @license
+ * Copyright (c) 2017 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {CRDTTypeRecord} from '../crdt/crdt.js';
+import {StorageCommunicationEndpointProvider, ProxyMessage, ProxyCallback, StorageCommunicationEndpoint} from './store.js';
+import {PropagatedException} from '../arc-exceptions.js';
+import {ChannelConstructor} from '../channel-constructor.js';
+import {StorageProxy} from './storage-proxy.js';
+import {Dictionary} from '../hot.js';
+import {Handle} from './handle.js';
+import {Type} from '../type.js';
+import {assert} from '../../platform/assert-web.js';
+
+export class BackingStorageProxy<T extends CRDTTypeRecord> implements StorageCommunicationEndpointProvider<CRDTTypeRecord> {
+  private storageProxies: Dictionary<StorageProxy<CRDTTypeRecord>> = {};
+  private callbacks: Dictionary<ProxyCallback<CRDTTypeRecord>> = {};
+  private storageEndpoint: StorageCommunicationEndpoint<T>;
+  private storageKey: string;
+  private type: Type;
+
+  constructor(storeProvider: StorageCommunicationEndpointProvider<T>, type: Type, storageKey: string) {
+    this.storageEndpoint = storeProvider.getStorageEndpoint(this);
+    this.storageEndpoint.setCallback(msg => this.onMessage(msg));
+    this.storageKey = storageKey;
+    this.type = type;
+  }
+  getStorageEndpoint(storageProxy: StorageProxy<CRDTTypeRecord>): StorageCommunicationEndpoint<CRDTTypeRecord> {
+    const backingStorageProxy = this;
+    const storageEndpoint = this.storageEndpoint;
+    const muxId = Object.keys(this.storageProxies).find(key => this.storageProxies[key] === storageProxy);
```

Comment:
Might be overkill, but consider using a BiMap for `this.storageProxies`?

https://github.com/PolymerLabs/arcs/blob/master/src/runtime/bimap.ts